### DISCUSSION TITLE
Detecting text changes in Rmd

### DISCUSSION THREAD

**Discussion by pat-s** [ropensci/tarchetypes#46] [Help]
Related #23 https://github.com/jdblischak/workflowr/issues/238

I was reading the issues related to {workflowr} + {targets} and currently using the following approach

```r
  tar_target(report_01,
    command = {
      !!tar_knitr_deps_expr("analysis/rep-01.Rmd")
      rmarkdown::render_site("analysis/rep-01.Rmd", quiet = TRUE)
      "docs/rep-01.html"
    },
    format = "file"
  ),
```

However, this approach does not watch for text-based changes in `rep-01.Rmd` as it only watches variables (which are extracted via `tar_knitr_deps_expr`).

IIRC in {drake} `knitr_in()` was able to also watch for text-based changes in Rmds. Is this somehow possible in {targets}?

**Comment by pat-s** [comment]
The following seems to work if one wants to watch for dependency **and** text based changes:

```r
  tar_target(report_01,
    command = {
      !!tar_knitr_deps_expr("analysis/rep-01.Rmd")
      suppressMessages(workflowr::wflow_build("analysis/rep-01.Rmd", verbose = FALSE))
      "analysis/rep-01.Rmd"
    },
    format = "file"
  ),
```

**Reply by wlandau** [reply]
Right, the R Markdown source needs to be part of the returned vector of paths. I also recommend returning the rendered HTML output so the target will automatically rerun if the output is corrupted. For example, `tarchetypes::tar_render(x, "report.Rmd")` returns "report.Rmd" and "report.html", as well as "report_files" if it exists.

**Reply by pat-s** [reply]
Ah! Ok, I see. Then how about this solution?

```r
tar_target(report_01,
  command = {
    !!tar_knitr_deps_expr("analysis/rep-01.Rmd")
    suppressMessages(workflowr::wflow_build("analysis/rep-01.Rmd", verbose = FALSE))
    c(
      "analysis/rep-01.Rmd",
      "docs/rep-01.html"
    )
  },
  format = "file"
)
```

This implies that the rendered output of `analysis/rep-01.Rmd"` is `"docs/rep-01.html"`.
The `analysis/` -> `docs/` relationship is the default in {workflowr}.

This solution 

- tracks dependency changes in `rep-01.Rmd`
- tracks text changes in `rep-01.Rmd`
- ensures that `rep-01.html` is present

Also cc @jdblischak

**Reply by wlandau** [reply]
Yeah, that will probably work. And your example target seems worthy of a target factory of its own, similar to the existing `tarchetypes::tar_render()`:

```r
tar_wflow_build(report_01, "analysis/rep-1.Rmd")
```

Though I am not sure whether it should live in `workflowr`, `tarchetypes`, or some other package.

**Reply by pat-s** [reply]
> Though I am not sure whether it should live in workflowr, tarchetypes, or some other package.

Me neither, given that I am not a maintainer of both :D 

Having it in {tarchetypes} would feel more natural to me given that there are so many additional `tar_` helpers. Also it would be easier for you to incorporate small changes if it would live in {tarchetypes} rather than in {workflowr}. But in the end this is a thing between you and @jdblischak :)

Also I've extended the above a bit more yesterday to also add custom deps to the target which do not appear in the Rmd itself:

```r
tar_target(report_01,
  command = {
    !!tar_knitr_deps_expr("analysis/rep-01.Rmd")
    suppressMessages(workflowr::wflow_build("analysis/rep-01.Rmd", verbose = FALSE))
   <CUSTOM DEP>
    c(
      "analysis/rep-01.Rmd",
      "docs/rep-01.html"
    )
  },
  format = "file"
)
```

This works well so far but meanwhile I fear that the whole command got quite complicated, especially for collaborators. Hence a dedicated wrapper would be highly appreciated because one cannot easily determine what role all lines play in the `command` block.

An example for a  `<CUSTOM DEP>` here is a CSS file target - i.e. if the CSS file was changed, I want the Rmd to get rebuilt. However the CSS is not explicitly loaded in the Rmd file as a target in a code block. Let me know if this is not good practice / some other way to do this :)

**Reply by wlandau** [reply]
> Having it in {tarchetypes} would feel more natural to me given that there are so many additional tar_ helpers. 

On reflection, my view on this is mixed because it may set a precedent for an unpredictably large `Suggests:` list. Maybe we are coming to a point where it makes sense to move all the R Markdown-related target factories to a new package of their own.

> An example for a <CUSTOM DEP> here is a CSS file target - i.e. if the CSS file was changed, I want the Rmd to get rebuilt. However the CSS is not explicitly loaded in the Rmd file as a target in a code block. Let me know if this is not good practice / some other way to do this :)

For CSS, you could append the path to the return value alongside `"analysis/rep-01.Rmd"` and `"docs/rep-01.html"`

**Reply by wlandau** [reply]
> On reflection, my view on this is mixed because it may set a precedent for an unpredictably large Suggests: list. 

The scope of `tarchetypes` is general-purpose interface extensions for `targets`. R Markdown itself is so general that I consider it to be in scope. But `workflowr`, `bookdown`, `blogdown`, `pkgdown`, `pagedown`, `xaringan`, etc. each get very specific, and like I said, the number of soft dependencies could easily creep up.

> Maybe we are coming to a point where it makes sense to move all the R Markdown-related target factories to a new package of their own.

A new `tarmarkdown` package for all things `targets` + R Markdown would require splitting `tarchetypes` into two pieces. This would be disruptive for users because they would need to load both `tarmarkdown` and `tarchetypes` to get the same functionality. This would also be a large maintenance burden for me because it would involve a lot of refactoring without a very clear gain. 

But if it were a package for just the out-of-scope stuff (`workflowr`, `bookdown`, `blogdown`, `pkgdown`, `pagedown`, `xaringan`, etc.) then that might be worth it. At the very least, it would not take us back a step. The scope would have to be clear enough though, and the package name should help with that (something like `tardown` or `targets.rmarkdown.extra`).

**Reply by petrbouchal** [reply]
Just to add one more to the list (sorry!): Rmarkdown websites. They are somewhat more complex (there is a setup file; much of the working dir is automatically copied to the output dir unless excluded) so handling them using the above approach gets quite verbose. But the relevant target factory would not introduce another dependency. 

The main value of such a target factory would be in 

- detecting all the Rmd files in scope for the website
- watching other `_site.yml` and files references in `_site.yml`, e.g. CSS and HTML includes
- watching the output directory as specified in `_site.yml`.

**Reply by wlandau** [reply]
R Markdown websites are a neat idea if we can reliably identify all the input and output files.

**Reply by ginolhac** [reply]
I have this very application working but am tracking things manually with `targets`. Would be awesome to have some helpers for those websites too!

**Reply by wlandau** [reply]
I had a look at what `render_site()` is doing, and I no longer think I have enough information to reliably create a target factory for the general case.  `render_site()` is not always explicit about the files it depends on, and do not think I can infer or replicate all of that faithfully. #49 may turn out to be an alternative, and I would like to see how that goes first.

**Reply by rgayler** [reply]
FWIW I have just posted https://github.com/jdblischak/workflowr/issues/238#issuecomment-849560808 over on {workflowr} showing how I am using {workflowr} + {targets}.

The key point is that I am only using {workflowr} notebooks to document the design of the core computational pipeline, which is managed by {targets}. So it is primarily a {targets} project, with {workflowr} only used to render the design documentation as a website.

The targets for the {workflowr} notebooks use `workflowr::wlfow_build()` to render the notebooks, but the decision to publish the notebooks strikes me as essentially manual, so I don't manage that aspect with {targets}. When I want to publish I manually invoke `workflowr::wflow_publish()`.

The target for one of these {workflowr} notebooks looks like:
```
tar_target(
        meta_notebook_1,
        command = {
            !! tar_knitr_deps_expr("meta_notebook_1.Rmd")
            list(core_fn_1, core_fn_2, core_fn_3)
            workflowr::wflow_build("meta_notebook_1.Rmd")
            "meta_notebook_1.Rmd""
        },
        format = "file"
    )
```
It checks for changes in the targets accessed by `tar_read()`, the core pipeline functions used in the Rmd file, and changes in the Rmd file itself. However, it does *not* check for changes in the rendered HTML file, because publishing the notebook would update the rendered file and invalidate the target.

`core_fn_1 ...` are functions to be used in the core pipeline that are developed and tested in the meta notebooks. So the function is a dependency for the relevant target in the core pipeline and also the meta notebook notebook where it is developed/used.
